packages/evm/src/opcodes/functions.ts

```diff
@@ -935,12 +935,15 @@ export const handlers: Map<number, OpHandler> = new Map([
         runState.stack.push(runState.cachedPushes[runState.programCounter])
         runState.programCounter += numToPush
       } else {
-        const loaded = bytesToBigInt(
-          runState.code.subarray(runState.programCounter, runState.programCounter + numToPush),
-        )
-        runState.programCounter += numToPush
-        runState.stack.push(loaded)
-      }
+        const slice = new Uint8Array(numToPush); 
+        const availableBytes = Math.min(numToPush, runState.code.length - runState.programCounter);
+    
+        slice.set(runState.code.subarray(runState.programCounter, runState.programCounter + availableBytes));
+    
+        const loaded = bytesToBigInt(slice);
+        runState.programCounter += numToPush;
+        runState.stack.push(loaded);
```

Comment:
These are some big changes to the PUSH opcode, which is ran a lot. We will need to keep performance in mind here, so I'll do some tests to see that the impact of this change is minimal.